src/plugins/upf/dpi.c

```diff
@@ -0,0 +1,1052 @@
+/*
+ * dpi.c - 3GPP TS 29.244 UPF DPI
+ *
+ * Copyright (c) 2017 Travelping GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define _LGPL_SOURCE            /* LGPL v3.0 is compatible with Apache 2.0 */
+
+#include <arpa/inet.h>
+
+#include <vppinfra/types.h>
+#include <vppinfra/vec.h>
+#include <vppinfra/pool.h>
+
+#include <hs.h>
+#include "upf/dpi.h"
+#include <upf/upf_pfcp.h>
+
+typedef struct {
+  regex_t *expressions;
+  u32 *ids;
+  u32 *flags;
+  hs_database_t *database;
+  hs_scratch_t *scratch;
+} upf_dpi_entry_t;
+
+typedef struct {
+  int res;
+  u32 id;
+} upf_dpi_cb_args_t;
+
+static upf_dpi_entry_t *upf_dpi_db = NULL;
+
+int
+upf_dpi_get_db_contents(u32 db_index, regex_t ** expressions, u32 ** ids)
+{
+  upf_dpi_entry_t *entry = NULL;
+
+  if (!db_index)
+    return -1;
+
+  if (pool_elts(upf_dpi_db) < db_index)
+    return -1;
+
+  db_index -= 1;
+
+  entry = pool_elt_at_index (upf_dpi_db, db_index);
+  if (!entry)
+    return -1;
+
+  *expressions = entry->expressions;
+  *ids = entry->ids;
+
+  return 0;
+}
+
+int
+upf_dpi_add_multi_regex(upf_dpi_args_t * args, u32 * db_index, u8 create)
```

Comment:
Could you elaborate why you are using an entry in the dpi_db per activated PDI ?

To me it seems that you are compiling a regex for every PDI that gets activate. Since we don't need to merge different applications into a big regex, that seems to be a lot overhead. Compiling the regex per application only when the application is initially loaded should suffice.

I would understand if you merged all DPI rules for a session into a big scan expression and the did some reprocessing to find out if a match was a false positive. But that seems not to be the case.